### ISSUE TITLE
A string type is always assumed when filtering by a field used as a partition

### ISSUE THREAD

**Issue by GregAru** [uber/petastorm#280]
I am trying to read data from a partitioned dataset in S3 where the partition column is defined as an integer:
`UnischemaField('partcol', np.int32, (), ScalarCodec(IntegerType()), False),`

The code below executes without exceptions but returns nothing:
```   
with make_reader(dataset_path, 
                     workers_count=1,
                     predicate=in_lambda(['partcol'], lambda partcol: (partcol == 20190102))
                     ) as reader:
```

The following code returns the expected data from the partition:
```
 with make_reader(dataset_path,
                     workers_count=1,
                     predicate=in_lambda(['partcol'], lambda partcol: (partcol == '20190102'))
                     ) as reader:
```

**Comment by selitvin** [comment]
Yep, we can address this since we know the type from Unischema.

**Comment by jsgoller1** [comment]
@GregAru - looking into this.

**Comment by jsgoller1** [comment]
@GregAru - while trying to reproduce the issue locally, I'm actually seeing the _opposite_ (though still incorrect) behavior, where a string passed as the partition column causes no data to be read, but an integer does:

```
    with make_reader(PETASTORM_DATASET_PATH,
                     workers_count=1,
                     predicate=in_lambda(['partcol'], lambda partcol: (partcol == '2'))
                     ) as reader:
        for schema_view in reader:
            print("This should return a schema_view:")
            print(schema_view)

    with make_reader(PETASTORM_DATASET_PATH,
                     workers_count=1,
                     predicate=in_lambda(['partcol'], lambda partcol: (partcol == 2))
                     ) as reader:
        for schema_view in reader:
            print("This should never be reached.")
            print(schema_view)
```
outputs the following:
```
(petastorm_venv3.6) root@computer:/petastorm# python test.py
This should never be reached.
HelloWorldSchema_view(id=2, partcol=2)
```
Is this at all consistent with what you are seeing (i.e. is the reported issue backwards)?

**Comment by GregAru** [comment]
@jsgoller1 - Not quite. In my case, the following does not return data:
```
     with make_reader(PETASTORM_DATASET_PATH,
                     workers_count=1,
                     predicate=in_lambda(['partcol'], lambda partcol: (partcol == 2))
                     ) as reader:
        for schema_view in reader:
            print("This should never be reached.")
            print(schema_view)
```

Please note that I am reading the dataset from S3 and the dataset is partitioned by `partcol `

**Comment by jsgoller1** [comment]
@GregAru  - thanks for the follow-up. I'm still having trouble reproducing this issue. I created a dataset with the following schema:

```HelloWorldSchema = Unischema('HelloWorldSchema', [
    UnischemaField('id', np.int32, (), ScalarCodec(IntegerType()), False),
    UnischemaField('partcol', np.int32, (), ScalarCodec(IntegerType()), False)
])```
and then executed my code above, which produces no output when the predicate compares `partcol` to a string, but correctly outputs the single schema view when it compares `partcol` to an int.

I have tried this both via python2.7 and python3.6, as well as locally and reading from S3. Are you able to share your dataset with us so that we can investigate the issue further?
